### ISSUE TITLE
Che workspace will not start if osio username is too big

### ISSUE THREAD

**Issue by rhopp** [openshiftio/openshift.io#2318]
Seems like a networking issue. Workspace pod seems to be successfully starter, but IDE times out and kills itself:
![image](https://user-images.githubusercontent.com/1215011/36472766-0a8eb496-16f3-11e8-9c62-3ed3e12798b9.png)

Che-server logs:
```
{"@timestamp":"2018-02-21T09:31:07.017+00:00","@version":1,"message":"Timeout. The Che server is unable to ping your workspace. This implies a network configuration issue, workspace boot failure, or an unusually slow workspace boot.","logger_name":"org.eclipse.che.api.workspace.server.WorkspaceManager","thread_name":"WorkspaceSharedPool-7","level":"ERROR","level_value":40000,"stack_trace":"org.eclipse.che.api.core.ServerException: Timeout. The Che server is unable to ping your workspace. This implies a network configuration issue, workspace boot failure, or an unusually slow workspace boot.\n\tat org.eclipse.che.api.agent.WsAgentLauncher.launch(WsAgentLauncher.java:153)\n\tat org.eclipse.che.api.workspace.server.WorkspaceRuntimes.launchAgents(WorkspaceRuntimes.java:681)\n\tat org.eclipse.che.api.workspace.server.WorkspaceRuntimes$MachineAgentsLauncher.started(WorkspaceRuntimes.java:1048)\n\tat org.eclipse.che.api.environment.server.CheEnvironmentEngine.startEnvironmentQueue(CheEnvironmentEngine.java:800)\n\tat org.eclipse.che.api.environment.server.CheEnvironmentEngine.start(CheEnvironmentEngine.java:263)\n\tat org.eclipse.che.api.workspace.server.WorkspaceRuntimes.startEnvironmentAndPublishEvents(WorkspaceRuntimes.java:694)\n\tat org.eclipse.che.api.workspace.server.WorkspaceRuntimes.access$100(WorkspaceRuntimes.java:103)\n\tat org.eclipse.che.api.workspace.server.WorkspaceRuntimes$StartTask.call(WorkspaceRuntimes.java:986)\n\tat org.eclipse.che.api.workspace.server.WorkspaceRuntimes$StartTask.call(WorkspaceRuntimes.java:948)\n\tat org.eclipse.che.commons.lang.concurrent.CopyThreadLocalCallable.call(CopyThreadLocalCallable.java:31)\n\tat java.util.concurrent.FutureTask.run(FutureTask.java:266)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\tat java.lang.Thread.run(Thread.java:748)\n"}
```

There is nothing suspicious in tenant namespace (no errors in events)

**Comment by eivantsov** [comment]
Routes may be unavailable even though ws agent is up

**Comment by rhopp** [comment]
I think I found a reason. My username is too long: 
![image](https://user-images.githubusercontent.com/1215011/36473238-414f053e-16f4-11e8-8b33-e40e11604c6b.png)

**Comment by ibuziuk** [comment]
@rhopp could you try how it works with different user ?

**Comment by ibuziuk** [comment]
@rhopp am I correct that the user name / namespace of the test account is `devtools-automated-tests-rh-che-us-east-2a-che` ? (46 chars)
Basically, there is a known limitation that max k8s names / labels length is 63 chars.

On OSIO custom router is used which completely ignores the 'host' field and assumes that routes host is supposed to have the following format  $NAME.$NAMESPACE where:
$NAME = route name
$NAMESPACE = namespace where route has been created.

In OpenShiftConnector host for route is never set and it is create automatically using  `$NAME.$NAMESPACE` pattern

Obviously there should be  username / namespace lengh limitation in place across osio in order to avoid such problems.

**Comment by ibuziuk** [comment]
This issue with username length (should be no more than 25 chars) restriction probably needs some PM attention cc: @slemeur @bmicklea @qodfathr

**Comment by eivantsov** [comment]
I think Che 6 handles that

**Comment by ibuziuk** [comment]
@eivantsov hmm.. wondering how Che6 could handle it since `$NAME.$NAMESPACE` for host is obligatory pattern for routes on osio?  (it is a must for route to have a namespace postfix)

**Comment by eivantsov** [comment]
@ibuziuk will check with Che 6! I know we had similar issue with pvcs

**Comment by ibuziuk** [comment]
@eivantsov yeah, but pvc is a a different story since there is no naming restriction from oso cluster point of view. I believe that in che6 route `name` generation has been changed, but `namespace` is still a part of host (in che6 host is not set explicitly while route creation and  is generated automatically using the same pattern `$NAME.$NAMESPACE`, otherwise it would not work). So, I bet there would be similar problem with che6.

**Comment by qodfathr** [comment]
Probably related; in fact, these may all, ultimately, be variations of the same bug: #2417, #2370

**Comment by ibuziuk** [comment]
@qodfathr yeah, this looks like the same issue. 

> As such,. in cases such as these, we need a safe way to translate between the naming systems -- even if that means using UUIDs in the remote systems and maintaining a mapping table of names to UUIDs. Yes, that makes it harder for a human to do the mapping manually (e.g. when in the OSO console) -- in instances like that, we should look for ways to mitigate (e.g. using UUIDs + partial name or adding more verbose text into Descriptions if available).

Do we have an issue for this ?

**Comment by l0rd** [comment]
@kpiwko @qodfathr my understanding was that to address this issue we would limit RHD username length. Am I correct?

**Comment by kpiwko** [comment]
As a workaround, yes, that's my understanding as well. As a real fix, there is more work to be done.

**Comment by l0rd** [comment]
Ok thanks @kpiwko. I remove the `priority/P0` label then. It doesn't look like it's required for GA. @qodfathr feel freed to add it back if you don't agree.

**Comment by bmicklea** [comment]
Has the RHD username limit been added?  I don't see it as a linked issue here so can't be sure...

**Comment by kpiwko** [comment]
No, the limit most likely has not been added, see from RHD team comment in the email thread:

> We do not have any support for this on our side currently. It means to develop something, and it will take some time definitely before we can introduce this.

So I'm re-adding P0

**Comment by l0rd** [comment]
## TL;DR

- With Che v6 we now support usernames of up to 45 chars (hence @rhopp original `devtools-automated-tests-rh-che-us-east-2a` should work now)
- To support unlimited usernames lenght: platform should change the way `-che` namespaces are named

## Details:

- Routes host names are generated automatically in the format `<route-name>-<namespace>.<suffix>` (e.g. `route5imc4gkr-mloriedo-che.8a09.starter-us-east-2.openshiftapps.com`)
- The part `<route-name>-<namespace>` is limited to 63 chars
- The part `<route-name>` is generated by Che and is currently 13 chars long (27 in Che v5)
- The part `<namespace>` is generated by Platform in the format `<username>-che` so currently username should be < 46 chars (32 in Che v5)

**Comment by joshuawilson** [comment]
Is this the same name used when the user signs up?
@alexeykazakov @sbose78 is this something we can control.

**Comment by bmicklea** [comment]
45 characters should be enough to prevent frequent issues. Removing P0 label but leaving as a SEV2.

**Comment by bmicklea** [comment]
@alexeykazakov if this is fixable before Summit without risking other issues go for it. Otherwise it can wait.

**Comment by alexeykazakov** [comment]
> Is this the same name used when the user signs up?

Yes.

> is this something we can control.

OSIO doesn't control usernames. RHD does. But OSIO could/should use something else for namespace names. Not usernames. This is something for our Tenant service.

> if this is fixable before Summit without risking other issues go for it. Otherwise it can wait.

No. It's not fixable before Summit. There are two options: Limit usernames in RHD (would take awhile to implement according to RHD guys) and it's not the best solution. Rather a workaround.
Or refactoring our tenant service to use something else. Not usernames.

**Comment by joshuawilson** [comment]
Is there any place we could give them a warning about the name length? Could we catch it in Che and let them know so they know what is happening?

**Comment by kpiwko** [comment]
I believe that what @joshuawilson has suggested is the best way how to "fix" that prior Summit - e.g. a big warning somewhere in the UI about username length.

**Comment by sbose78** [comment]
Can we ensure that we don't approve/provision users with long usernames? That should be available in the registration app dashboard.

**Comment by qodfathr** [comment]
@sbose78 exactly what would we do with those users? They'd be stuck in limbo, with zero information as to why.

**Comment by alexeykazakov** [comment]
@qodfathr correct. If we just reject user's approval/provisioning if the username is longer than 46 then the user will stuck in the waiting list without any information as to why.

**Comment by sbose78** [comment]
> exactly what would we do with those users?

Communicate to these users that their username is too long and that need to re-register with a shorter one. That's better than them getting those users access to OpenShift.io and then failing to even get the golden path working.

**Comment by bmicklea** [comment]
Who would do that, and where would we get the emails?  Is it automated?

**Comment by alexeykazakov** [comment]
Please note that steps 3-5 are manual steps. It's definitely can't be a long term solution. I even doubt it's a good short term solution but this is what we could do now.

**Comment by alexeykazakov** [comment]
Define "we" :)
RHD lets people set their usernames. And yes, in same cases RHD doesn't let use certain usernames (already used for example). As said above the proper place for such limitation would be the RHD registration form. But it will requires changes on RHD site which is not trivial if I understood @velias correctly.

@velias could you please estimate such work on the RHD site? @qodfathr maybe you can work with RHD PMs to help prioritize this task.

**Comment by velias** [comment]
RHD currently allows max 45 chars in the username. We introduced this limit few weeks ago. But some legacy accounts may exist with longer username.

**Comment by bmicklea** [comment]
Thanks @velias. Given the earlier comment from @l0rd about Che support up to 45 characters in the name it sounds like this issue is only going to affect any users with larger user names already in the system.

Closing as fixed. If I'm mistaken re-open.
